pkg/kv/kvserver/client_raft_test.go

```diff
@@ -5921,3 +5921,393 @@ func TestRaftSnapshotsWithMVCCRangeKeysEverywhere(t *testing.T) {
+							},
+						},
+					},
+				})
+				defer tc.Stopper().Stop(ctx)
+
+				logStatus := func(s *raft.Status) {
+					t.Helper()
+					require.NotNil(t, s)
+					t.Logf("n%d %s at term=%d commit=%d", s.ID, s.RaftState, s.Term, s.Commit)
+				}
+
+				// Create a range, upreplicate it, and replicate a write.
+				sender := tc.GetFirstStoreFromServer(t, 0).TestSender()
+				key := tc.ScratchRange(t)
+				desc := tc.AddVotersOrFatal(t, key, tc.Targets(1, 2)...)
+				rangeID = desc.RangeID
+
+				_, pErr := kv.SendWrapped(ctx, sender, incrementArgs(key, 1))
+				require.NoError(t, pErr.GoError())
+				tc.WaitForValues(t, key, []int64{1, 1, 1})
+
+				repl1, err := tc.GetFirstStoreFromServer(t, 0).GetReplica(rangeID)
+				require.NoError(t, err)
+				repl3, err := tc.GetFirstStoreFromServer(t, 2).GetReplica(rangeID)
+				require.NoError(t, err)
+
+				// Configure the partition, but don't activate it yet.
+				if partial {
+					// Partition n3 away from n1, in both directions.
+					dropRaftMessagesFrom(t, tc.Servers[0], rangeID, []roachpb.ReplicaID{3}, &partitioned)
+					dropRaftMessagesFrom(t, tc.Servers[2], rangeID, []roachpb.ReplicaID{1}, &partitioned)
+				} else {
+					// Partition n3 away from both of n1 and n2, in both directions.
+					dropRaftMessagesFrom(t, tc.Servers[0], rangeID, []roachpb.ReplicaID{3}, &partitioned)
+					dropRaftMessagesFrom(t, tc.Servers[1], rangeID, []roachpb.ReplicaID{3}, &partitioned)
+					dropRaftMessagesFrom(t, tc.Servers[2], rangeID, []roachpb.ReplicaID{1, 2}, &partitioned)
+				}
+
+				// Make sure the lease is on n1 and that everyone has applied it.
+				tc.TransferRangeLeaseOrFatal(t, desc, tc.Target(0))
+				_, pErr = kv.SendWrapped(ctx, sender, incrementArgs(key, 1))
+				require.NoError(t, pErr.GoError())
+				tc.WaitForValues(t, key, []int64{2, 2, 2})
+				t.Logf("n1 has lease")
+
+				// Wait for the range to quiesce, if enabled.
+				if quiesce {
+					require.Eventually(t, repl3.IsQuiescent, 10*time.Second, 100*time.Millisecond)
+					t.Logf("n3 quiesced")
+				} else {
+					require.False(t, repl3.IsQuiescent())
+					t.Logf("n3 not quiesced")
+				}
+
+				// Partition n3.
+				partitioned.Store(true)
+				t.Logf("n3 partitioned")
+
+				// Fetch the leader's initial status.
+				initialStatus := repl1.RaftStatus()
+				require.Equal(t, raft.StateLeader, initialStatus.RaftState)
+				logStatus(initialStatus)
+
+				// Unquiesce n3 if necessary.
+				if quiesce {
+					require.True(t, repl3.MaybeUnquiesce())
+					t.Logf("n3 unquiesced")
+				} else {
+					require.False(t, repl3.IsQuiescent())
+					t.Logf("n3 not quiesced")
+				}
+
+				// Wait for the follower to become a candidate.
+				require.Eventually(t, func() bool {
+					status := repl3.RaftStatus()
+					logStatus(status)
+					return status.RaftState == raft.StatePreCandidate
+				}, 10*time.Second, 500*time.Millisecond)
+				t.Logf("n3 became pre-candidate")
+
+				// The candidate shouldn't change state for 3 seconds.
```

Comment:
Why 3 seconds? If you put 10 seconds here, would it still stay in the same state? Just checking that there isn't a race here with some fixed duration.

Reply:
Correct. We just have to wait "long enough" that the replica won't transition into Candidate or Follower, either because of any received messages or because of an election timeout. Updated the comment to clarify this.